pkg/provider/azure_instances.go

```diff
@@ -385,22 +385,10 @@ func (az *Cloud) getLocalInstanceProviderID(metadata *InstanceMetadata, nodeName
 	resourceGroup := strings.ToLower(metadata.Compute.ResourceGroup)
 	subscriptionID := strings.ToLower(metadata.Compute.SubscriptionID)
 
-	// Compose instanceID based on nodeName for standard instance.
-	if metadata.Compute.VMScaleSetName == "" {
-		return az.getStandardMachineID(subscriptionID, resourceGroup, nodeName), nil
-	}
+	providerID := strings.Replace(metadata.Compute.ResourceID, metadata.Compute.SubscriptionID, subscriptionID, -1)
+	providerID = strings.Replace(providerID, metadata.Compute.ResourceGroup, resourceGroup, -1)
 
-	// Get scale set name and instanceID from vmName for vmss.
-	ssName, instanceID, err := extractVmssVMName(metadata.Compute.Name)
-	if err != nil {
-		if errors.Is(err, ErrorNotVmssInstance) {
-			// Compose machineID for standard Node.
-			return az.getStandardMachineID(subscriptionID, resourceGroup, nodeName), nil
-		}
-		return "", err
-	}
-	// Compose instanceID based on ssName and instanceID for vmss instance.
-	return az.getVmssMachineID(subscriptionID, resourceGroup, ssName, instanceID), nil
+	return providerID, nil
```

Comment:
In case there are something wrong with IMDS responses, could you check whether "resourceID" is not empty here?

If it is empty, then IMDS cache needs to be refreshed, refer

https://github.com/kubernetes-sigs/cloud-provider-azure/blob/35f325bc6b9e2539e2798da062b675b7aab75e66/pkg/provider/azure_instances.go#L152-L153

Reply:
I see, thanks for sharing this info. I just changed the code to handle the empty resourceID case